pkg/kubelet/status/status_manager.go

```diff
@@ -682,34 +682,47 @@ type SidecarsStatus struct {
 }
 
 // GetSidecarsStatus returns the SidecarsStatus for the given pod
-func GetSidecarsStatus(pod *v1.Pod) SidecarsStatus {
+func GetSidecarsStatus(pod *v1.Pod, caller string) SidecarsStatus {
 	if pod == nil {
-		klog.Infof("Pod was nil, returning empty sidecar status")
+		klog.Infof("Pod was nil, returning empty sidecar status (%s)", caller)
 		return SidecarsStatus{}
 	}
-	if pod.Spec.Containers == nil || pod.Status.ContainerStatuses == nil {
-		klog.Infof("Pod Containers or Container status was nil, returning empty sidecar status")
+	if pod.Spec.Containers == nil {
+		klog.Infof("Pod Containers was nil, returning empty sidecar status (%s)", caller)
 		return SidecarsStatus{}
 	}
+	if pod.Status.ContainerStatuses == nil {
+		klog.Infof("Pod Container status was nil, doing best effort using spec (%s)", caller)
+	}
 	sidecarsStatus := SidecarsStatus{SidecarsPresent: false, SidecarsReady: true, ContainersWaiting: false}
 	for _, container := range pod.Spec.Containers {
+		foundStatus := false
+		isSidecar := false
+		if pod.Annotations[fmt.Sprintf("sidecars.lyft.net/container-lifecycle-%s", container.Name)] == "Sidecar" {
+			isSidecar = true
+			sidecarsStatus.SidecarsPresent = true
+		}
 		for _, status := range pod.Status.ContainerStatuses {
+			foundStatus = true
 			if status.Name == container.Name {
-				if pod.Annotations[fmt.Sprintf("sidecars.lyft.net/container-lifecycle-%s", container.Name)] == "Sidecar" {
-					sidecarsStatus.SidecarsPresent = true
+				if isSidecar {
 					if !status.Ready {
-						klog.Infof("Pod %s: %s: sidecar not ready", format.Pod(pod), container.Name)
+						klog.Infof("Pod %s: %s: sidecar not ready (%s)", format.Pod(pod), container.Name, caller)
 						sidecarsStatus.SidecarsReady = false
 					} else {
-						klog.Infof("Pod %s: %s: sidecar is ready", format.Pod(pod), container.Name)
+						klog.Infof("Pod %s: %s: sidecar is ready (%s)", format.Pod(pod), container.Name, caller)
 					}
 				} else if status.State.Waiting != nil {
 					// check if non-sidecars have started
-					klog.Infof("Pod: %s: %s: non-sidecar waiting", format.Pod(pod), container.Name)
+					klog.Infof("Pod: %s: %s: non-sidecar waiting (%s)", format.Pod(pod), container.Name, caller)
 					sidecarsStatus.ContainersWaiting = true
 				}
 			}
 		}
+		if !foundStatus && isSidecar {
```

Comment:
This could move up and bail early.

Reply:
I've run through this in my mind a few different ways and nothing stands out that seems good.

I have however added a `break` out of the loop in the `if` above to bail early once a status is found for a given container.